### PR TITLE
add basic auth option to apiserver in local cluster

### LINKED ISSUE #40192: local-up-cluster causes error: You must be logged in to the server (the server has asked for the client to provide credentials)
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

Searched for:

 is:issue is:open local-up-cluster  error: You must be logged in to the server (the server has asked for the client to provide credentials)  

I did not find anything relevant. 

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): 
BUG REPORT 

**Kubernetes version** (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.0.3557+05c2a0825e39be-dirty", GitCommit:"05c2a0825e39beee3de7e7bb1b49d8083ef90d65", GitTreeState:"dirty", BuildDate:"2017-01-20T01:19:49Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.0.3557+05c2a0825e39be", GitCommit:"05c2a0825e39beee3de7e7bb1b49d8083ef90d65", GitTreeState:"clean", BuildDate:"2017-01-19T18:10:47Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
Local Laptop  Mac OS 10.11.06
- **Cloud provider or hardware configuration**:
VirtualBox Vagrant 
- **OS** (e.g. from /etc/os-release):
centos7.2 in vagrant vm
- **Kernel** (e.g. `uname -a`):
Linux k8s-env 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**:
- **Others**:

**What happened**:

After running hack-local-up-cluster.sh the script prints the following 

```
  export KUBERNETES_PROVIDER=local

  cluster/kubectl.sh config set-cluster local --server=https://localhost:6443 --certificate-authority=/var/run/kubernetes/apiserver.crt
  cluster/kubectl.sh config set-credentials myself --username=admin --password=admin
  cluster/kubectl.sh config set-context local --cluster=local --user=myself
  cluster/kubectl.sh config use-context local
  cluster/kubectl.sh
```

Running these results in an authentication error 

```
[vagrant@k8s-env kubernetes]$
[vagrant@k8s-env kubernetes]$   export KUBERNETES_PROVIDER=local
[vagrant@k8s-env kubernetes]$
[vagrant@k8s-env kubernetes]$   cluster/kubectl.sh config set-cluster local --server=https://localhost:6443 --certificate-authority=/var/run/kubernetes/apiserver.crt
Cluster "local" set.
[vagrant@k8s-env kubernetes]$   cluster/kubectl.sh config set-credentials myself --username=admin --password=admin
User "myself" set.
[vagrant@k8s-env kubernetes]$   cluster/kubectl.sh config set-context local --cluster=local --user=myself
Context "local" set.
[vagrant@k8s-env kubernetes]$   cluster/kubectl.sh config use-context local
Switched to context "local".
[vagrant@k8s-env kubernetes]$   cluster/kubectl.sh cluster-info
error: You must be logged in to the server (the server has asked for the client to provide credentials)
````

**What you expected to happen**:

I would have expected to authenticate with the api server. 

**How to reproduce it** (as minimally and precisely as possible):

Go over the same steps 

**Anything else do we need to know**:

In the problematic case, the ~/.kube/config file looks like the following

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /var/run/kubernetes/apiserver.crt
    server: https://localhost:6443
  name: local
contexts:
- context:
    cluster: local
    user: myself
  name: local
current-context: local
kind: Config
preferences: {}
users:
- name: myself
  user:
    password: admin
    username: admin
```

Setting  ALLOW_ANY_TOKEN is one workaround. 
```
export ALLOW_ANY_TOKEN=true
```
With this env var the cluster/kubectl.sh command can talk to the api server. 

I recommend to start from a clean slate to reproduce the issue. 

Once the authentication succeeds with setting the ALLOW_ANY_TOKEN=true, I have tried to go back and restart the local-up-cluster with  ALLOW_ANY_TOKEN=false again. In the second attempt the authentication SUCCEEDED. So I see the problem only at the first attempt.

### PR DESCRIPTION
**Pull request by xilabao**
Add this can fix #40192 @liggitt

### CODE REVIEW

**Comment by liggitt** [comment]
I'm wondering if the output should change to reference the generated kubeconfig that already contains the correct info, rather than making them make their own

**Comment by xilabao** [comment]
`/var/run/kubernetes/admin.kubeconfig` is a option. I think let people to run `kubectl --kubeconfig=...` is not so good. We can update the local kubeconfig instead the output.

**Comment by xilabao** [comment]
@liggitt  The output updated. How about this? 
```
To start using your cluster, you can add kubeconfig as '--kubeconfig=/var/run/kubernetes/admin.kubeconfig' in kubectl.
Alternatively, you can open up another terminal/tab and run (For simplicity, LOCAL_KUBECONFIG_PATH can update the local kubeconfig):

  export KUBERNETES_PROVIDER=local

  cluster/kubectl.sh config set-cluster local --server=https://localhost:6443 --certificate-authority=/var/run/kubernetes/apiserver.crt
  cluster/kubectl.sh config set-credentials myself --username=admin --password=admin
  cluster/kubectl.sh config set-context local --cluster=local --user=myself
  cluster/kubectl.sh config use-context local
  cluster/kubectl.sh
```

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -665,8 +688,15 @@ fi
...
-  cat <<EOF
-To start using your cluster, open up another terminal/tab and run:
+  if [[ -n "${LOCAL_KUBECONFIG_PATH}" ]]; then
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config set-cluster local --server=https://${API_HOST}:${API_SECURE_PORT} --certificate-authority=${ROOT_CA_FILE}
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config set-credentials myself ${AUTH_ARGS}
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config set-context local --cluster=local --user=myself
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config use-context local
+  else
+    cat <<EOF
+To start using your cluster, you can add kubeconfig as '--kubeconfig=/var/run/kubernetes/admin.kubeconfig' in kubectl.
```
Comment (euank) [inline hack/local-up-cluster.sh]: I would rather have the instructions be:

 ```
export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
./cluster/kubectl.sh
```

Comment (xilabao) [inline hack/local-up-cluster.sh]: I forgot about that. Thanks. :)

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -665,8 +688,19 @@ fi
...
+  if [[ -n "${LOCAL_KUBECONFIG_PATH}" ]]; then
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config set-cluster local --server=https://${API_HOST}:${API_SECURE_PORT} --certificate-authority=${ROOT_CA_FILE}
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config set-credentials myself ${AUTH_ARGS}
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config set-context local --cluster=local --user=myself
+    ${KUBECTL} --kubeconfig=${LOCAL_KUBECONFIG_PATH} config use-context local
+  else
+    cat <<EOF
+To start using your cluster, you can open up another terminal/tab and run:
+
+  export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig
```
Comment (liggitt) [inline hack/local-up-cluster.sh]: use ${CERT_DIR}, don't hard-code /var/run/...

@deads2k you have opinions about how we should tell people to use this kubeconfig?

Comment (deads2k) [inline hack/local-up-cluster.sh]: > @deads2k you have opinions about how we should tell people to use this kubeconfig?

readonly.  Let's create a second skeleton for a user.

**Comment by xilabao** [comment]
fixed. @liggitt

**Comment by eparis** [comment]
@k8s-bot ok to test

**Comment by hongchaodeng** [comment]
I wasn't able to get it worked:
```
Waiting for apiserver to come up
!!! [0127 03:31:52] Timed out waiting for apiserver:  to answer at http://localhost:8080/version; tried 10 waiting 1 between each
Cleaning up...
```

@xilabao How did you run this?

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -665,8 +688,19 @@ fi
 
 if [[ "${START_MODE}" != "kubeletonly" ]]; then
   echo
-  cat <<EOF
-To start using your cluster, open up another terminal/tab and run:
+  if [[ -n "${LOCAL_KUBECONFIG_PATH}" ]]; then
```
Comment (liggitt) [inline hack/local-up-cluster.sh]: I don't think the complexity LOCAL_KUBECONFIG_PATH adds is worth it.

**Comment by xilabao** [comment]
remove LOCAL_KUBECONFIG_PATH @liggitt PTAL

**Comment by xilabao** [comment]
@hongchaodeng  Do you still can not get it worked?  I just run `hack/local-up-cluster.sh`

**Comment by hongchaodeng** [comment]
@xilabao No.

**Comment by hongchaodeng** [comment]
Can you share how you run it? Incl. etcd versions

**Comment by xilabao** [comment]
@hongchaodeng 
```
etcd --version
etcd Version: 3.0.14
Git SHA: 8a37349
Go Version: go1.6.3
Go OS/Arch: linux/amd64
```

```
./cluster/kubectl.sh version
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.1.199+5ea6fc3f33a227", GitCommit:"5ea6fc3f33a2272a0635f08b08c05b20fca9fca5", GitTreeState:"clean", BuildDate:"2017-02-02T02:44:51Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
error: You must be logged in to the server (the server has asked for the client to provide credentials (get .meta.k8s.io))
```

**Comment by hongchaodeng** [comment]
How do you start the cluster?
I'm fetching exactly your commit 6069cf4
Then run:
```
$ ./hack/local-up-cluster.sh
...
2017/02/02 04:05:22 [INFO] generate received request
2017/02/02 04:05:22 [INFO] received CSR
2017/02/02 04:05:22 [INFO] generating key: rsa-2048
2017/02/02 04:05:22 [INFO] encoded CSR
2017/02/02 04:05:22 [INFO] signed certificate with serial number 415626845969731860716086602635216065429307606679
Waiting for apiserver to come up
!!! [0202 04:05:32] Timed out waiting for apiserver:  to answer at http://localhost:8080/version; tried 10 waiting 1 between each
Cleaning up...
```

**Comment by xilabao** [comment]
@hongchaodeng  The commit https://github.com/kubernetes/kubernetes/commit/6069cf4451cb457a2e5c72582116ef1324a33b88  guides people to set the kubeconfig if the cluster has started. The error seems that the apiserver doesn't start. Do you in the lastest master branch?

**Comment by hongchaodeng** [comment]
>  Do you in the lastest master branch?

It's the branch that you pushed to this PR:
```
("upstream" indicates k8s git repo)
$ git fetch upstream pull/40193/head:local-fix
$ git checkout local-fix
```

**Comment by xilabao** [comment]
@hongchaodeng  Is `/var/run/kubernetes/basic_auth.csv` exist? Do you have permission to create `/var/run/kubernetes/basic_auth.csv`?

**Comment by hongchaodeng** [comment]
@xilabao 
> Do you have permission to create /var/run/kubernetes/basic_auth.csv

Right. Permission problem. I got it worked finally. Thanks for helping out :)

**Comment by kargakis** [comment]
Not sure when local-up-cluster was working ootb last time on master. Can we get this merged?

**Comment by liggitt** [comment]
I will look today. I'm still not sure we want basic auth,  but the documentation around the generated kubeconfig file would be good to add

**Comment by liggitt** [comment]
Using the generated file, or running directly against the unsecured port works out of the box today. The instructions for using basic auth are just incorrect

File: hack/local-up-cluster.sh [hunk 4]
```diff
@@ -447,7 +469,7 @@ function start_apiserver {
             fi
         else
             # default to use basic authentication
-            AUTH_ARGS="--username=admin --password=admin"
+            AUTH_ARGS="--username=${KUBE_USER} --password=${KUBE_PASSWORD}"
```
Comment (liggitt) [inline hack/local-up-cluster.sh]: I'd rather see this change to reference the generated admin cert/key (going back to something closer to before https://github.com/kubernetes/kubernetes/commit/72e0e91b5edb961cd22e2cdbce5d02300a754402#diff-ed3df710e9af7cd30a185896a60897d9L713)

```
AUTH_ARGS="--client-key=${CERT_DIR}/client-admin.key --client-certificate=${CERT_DIR}/client-admin.crt"
```

**Comment by liggitt** [comment]
local-up-cluster.sh has never supported basic auth as far as I can see (history doesn't include any basic-auth-file). 72e0e91b5edb961cd22e2cdbce5d02300a754402 added instructions to use `--username=admin --password=admin` but did not actually do anything to make that work. I'd prefer we use the admin client cert/key already in place, and just update the instructions to be correct

**Comment by xilabao** [comment]
@liggitt 
```
hack/local-up-cluster.sh
WARNING : This script MAY be run as root for docker socket / iptables functionality; if failures occur, retry as root.
....
```
hack/local-up-cluster.sh suggest to run as root.  If run as root, may meet the problem below.
```
./cluster/kubectl.sh get no
error: unable to read client-key /var/run/kubernetes/client-admin.key for myself due to open /var/run/kubernetes/client-admin.key: permission denied
```
Should we ignore it?

**Comment by liggitt** [comment]
options:
* include chmod instructions if started with sudo
* let the user `sudo su -` to use a cluster they started as root
* chmod the kubeconfig/cert/key automatically (insecure, but local access is already unsecured over :8080)

@sttts, preference?

**Comment by sttts** [comment]
@liggitt why does one start a cluster as root and then expect it's usable by non-root?

**Comment by sttts** [comment]
I am not a fan of enabling basic auth. I agree with @liggitt that updating the instructions is the way to go.

**Comment by xilabao** [comment]
@liggitt PTAL

File: hack/local-up-cluster.sh [hunk 5]
```diff
@@ -436,20 +435,6 @@ function start_apiserver {
...
-        if [[ "${ALLOW_ANY_TOKEN}" = true  ]]; then
-            # use token authentication
-            if [[ -n "${KUBECONFIG_TOKEN}"  ]]; then
-                AUTH_ARGS="--token=${KUBECONFIG_TOKEN}"
-            else
-                AUTH_ARGS="--token=system:admin/system:masters"
-            fi
-        else
-            # default to use basic authentication
-            AUTH_ARGS="--username=admin --password=admin"
```
Comment (liggitt) [inline hack/local-up-cluster.sh]: Keeping anytoken doesn't bother me (cc @deads2k), can we just change the basic auth back to client cert auth, and include the admin.kubeconfig info? That seems the minimal change.

Comment (xilabao) [inline hack/local-up-cluster.sh]: Done. PTAL

File: hack/local-up-cluster.sh [hunk 6]
```diff
@@ -673,10 +673,20 @@ elif [[ "${START_MODE}" == "nokubelet" ]]; then
 fi
 
 if [[ "${START_MODE}" != "kubeletonly" ]]; then
+  USE_SUDO=""
+  if [ "$(id -u)" == "0"  ]; then
+      USE_SUDO="sudo su -"
```
Comment (liggitt) [inline hack/local-up-cluster.sh]: with @sttts change, I don't think sudo is required any more

**Comment by liggitt** [comment]
fix rebase, verify sudo isn't required any more, then LGTM

**Comment by xilabao** [comment]
fixed. @liggitt

File: hack/local-up-cluster.sh [hunk 7]
```diff
@@ -688,7 +688,12 @@ fi
 if [[ "${START_MODE}" != "kubeletonly" ]]; then
   echo
   cat <<EOF
-To start using your cluster, open up another terminal/tab and run:
+To start using your cluster, you can open up another terminal/tab and run:
+
+  export KUBECONFIG=${CERT_DIR}/admin.kubeconfig
```
Comment (sttts) [inline hack/local-up-cluster.sh]: 👍

**Comment by MHBauer** [comment]
I think this also fixed #40459.

**Pull request merged**